### PR TITLE
examples: Add dynamic configuration (control plane) sandbox

### PR DESCRIPTION
**Pull request by phlax**
Signed-off-by: Ryan Northey <ryan@synca.io>

Commit Message: examples: Add control plane dynamic configuration sandbox
Additional Description:

Add example of using dynamic configuration with go-control-plane

Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]

### CODE REVIEW

File: examples/dynamic-configuration/envoy.yaml
Comment (phlax) [inline examples/dynamic-configuration/envoy.yaml:15]: not sure why this was necessary, but it wouldnt update without it

Comment (dio) [reply examples/dynamic-configuration/envoy.yaml:15]: Probably it is related to this: https://github.com/envoyproxy/envoy/issues/9682?

File: docs/root/start/sandboxes/_include/dynamic-configuration/response-config-cluster.json
Comment (phlax) [inline docs/root/start/sandboxes/_include/dynamic-configuration/response-config-cluster.json:4]: this is the wrong response - update to v3

File: examples/dynamic-configuration/verify.sh
Comment (phlax) [inline examples/dynamic-configuration/verify.sh:59]: use `wait_for`

File: docs/root/start/sandboxes/dynamic-configuration.rst
Comment (phlax) [inline docs/root/start/sandboxes/dynamic-configuration.rst:10]: im thinking  to move this section to the quick-start page

File: docs/root/start/sandboxes/dynamic-configuration.rst
Comment (phlax) [inline docs/root/start/sandboxes/dynamic-configuration.rst:169]: this seems like an Envoy bug (maybe im wrong)

basically if you configure a backend  - in this case `service` and then change the configuration dynamically - ie to `service2`, Envoy starts using the correct backend (ie ip) but addresses the new backend with the old backend host name (`service1`)

@dio any ideas about this ?

File: docs/root/start/sandboxes/dynamic-configuration.rst
Comment (phlax) [reply docs/root/start/sandboxes/dynamic-configuration.rst]: removed for now - ill pr to  update quick-start once this lands

File: examples/dynamic-configuration/README.md
Comment (phlax) [inline examples/dynamic-configuration/README.md:2]: fix this

File: docs/root/start/sandboxes/_include/dynamic-configuration/response-config-cluster.json
Comment (phlax) [reply docs/root/start/sandboxes/_include/dynamic-configuration/response-config-cluster.json:4]: hmm - seems like this is the response you get back from the go-control-plane

Comment (phlax) [reply docs/root/start/sandboxes/_include/dynamic-configuration/response-config-cluster.json:4]: i guess this must be what is configured in `static_resources` - im surprised its not an `api.v3.Cluster`

File: docs/root/start/sandboxes/dynamic-configuration.rst
Comment (phlax) [reply docs/root/start/sandboxes/dynamic-configuration.rst:183]: im wondering if this is related to #13659

**Comment by phlax** [comment]
this PR is blocked by #13749

**Comment by mattklein123** [comment]
Do you want to land this independently of the other getting started docs? Either way lmk when this is ready for review and not draft.

/wait

**Comment by phlax** [comment]
this should be ready to land almost straight away, once ive resolved the conflicts, and gone through again

File: docs/root/start/sandboxes/dynamic-configuration.rst
Comment (phlax) [reply docs/root/start/sandboxes/dynamic-configuration.rst:183]: @mattklein123 any ideas about this ?

im wondering if i should open a ticket for it - or if its just a dupe of the ticket i posted

File: examples/dynamic-configuration/envoy.yaml
Comment (phlax) [reply examples/dynamic-configuration/envoy.yaml]: cool, ill watch that ticket and update if possible/necessary after

File: docs/root/start/sandboxes/_include/dynamic-configuration/response-config-cluster.json
Comment (mattklein123) [reply docs/root/start/sandboxes/_include/dynamic-configuration/response-config-cluster.json:4]: Yeah I would update to v3 if possible. cc @htuch @kyessenov for guidance here.

File: docs/root/start/sandboxes/dynamic-configuration.rst
Comment (mattklein123) [reply docs/root/start/sandboxes/dynamic-configuration.rst:183]: How is host being set? Are you using auto host rewrite?  I'm confused who is setting host here. Also, it might be better to use static clusters for this example vs. DNS clusters?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks very cool to have this kind of example.

/wait

File: docs/root/start/sandboxes/dynamic-configuration.rst
Comment (phlax) [reply docs/root/start/sandboxes/dynamic-configuration.rst:183]: How is host being set? 

afaict host is being set by Envoy - its  the only thing that knows the names `service1/2` (apart from docker)

Are you using auto host rewrite? 

not sure - defaults from go-control-plane. You can  see the `cluster` config here 

https://storage.googleapis.com/envoy-pr/13746/docs/start/sandboxes/dynamic-configuration.html#step-4-check-initial-config-and-web-response

altho, im wondering if this is in `listeners`

i tried to override it by setting the host header but didnt work -  got same result

> Also, it might be better to use static clusters for this example vs. DNS clusters?

you mean with fixed IP ?

i think its good to show it like this with dns coming from docker etc (thats how most others do i think) - also, if it is fixed by ip we would have to fix that in docker (maybe i misunderstand)

either way - afaict this behaviour/bug is wrong/undesirable even if we fix the example

im wondering if its the way go-control-plane is setting/updating the config - ill dump more and try to see...

Comment (phlax) [reply docs/root/start/sandboxes/dynamic-configuration.rst]: after further investigation this seems to be an issue with go-control-plane - at least with the example setup

i ran through the sandbox, and then dumped config...

```console
$ curl -s http://localhost:19000/config_dump  | jq '.configs[4]' 
{
  "@type": "type.googleapis.com/envoy.admin.v3.RoutesConfigDump",
  "dynamic_route_configs": [
    {
      "version_info": "1",
      "route_config": {
        "@type": "type.googleapis.com/envoy.config.route.v3.RouteConfiguration",
        "name": "local_route",
        "virtual_hosts": [
          {
            "name": "local_service",
            "domains": [
              "*"
            ],
            "routes": [
              {
                "match": {
                  "prefix": "/"
                },
                "route": {
                  "cluster": "example_proxy_cluster",
                  "host_rewrite_literal": "service1"
                }
              }
            ]
          }
        ]
      },
      "last_updated": "2020-10-27T19:06:56.600Z"
    }
  ]
}
```

it seems like updating something in go-control-plane is not updating `host_rewrite_literal` when the upstream gets changed

its not an issue with envoy - so sorry for noise

Comment (phlax) [reply docs/root/start/sandboxes/dynamic-configuration.rst]: it seems like its not updating routes config after the first snapshot, as that is still showing `version=1` even though the rest of the config has `2`

Comment (phlax) [reply docs/root/start/sandboxes/dynamic-configuration.rst]: im gonna set this as resolved - if its an issue, its outside the scope of this PR

File: docs/root/start/sandboxes/_include/dynamic-configuration/response-config-cluster.json
Comment (phlax) [reply docs/root/start/sandboxes/_include/dynamic-configuration/response-config-cluster.json]: this seems like an issue with go-control-plane - (at least the example i based this on)

Comment (phlax) [reply docs/root/start/sandboxes/_include/dynamic-configuration/response-config-cluster.json]: which is odd because checking here https://github.com/envoyproxy/go-control-plane/tree/master/internal/example i cant see a `v2` anywhere - everything is set to v3

also checking back through this sandbox - all of the config suggests v3  - so im a bit lost as to cause

**Comment by phlax** [comment]
docs are rendered here

https://storage.googleapis.com/envoy-pr/13746/docs/start/sandboxes/dynamic-configuration-control-plane.html

File: docs/root/start/sandboxes/dynamic-configuration-control-plane.rst
Comment (phlax) [inline docs/root/start/sandboxes/dynamic-configuration-control-plane.rst:151]: add `lineno-start`

File: docs/root/start/sandboxes/_include/dynamic-configuration/response-config-cluster.json
Comment (phlax) [reply docs/root/start/sandboxes/_include/dynamic-configuration/response-config-cluster.json]: looking at this further its nothing to do with go-control-plane as this is Envoy's configuration for g-c-p

im wondering if there is anything i can change to this config:

```
dynamic_resources:
  ads_config:
    api_type: GRPC
    transport_api_version: V3
    grpc_services:
      - envoy_grpc:
          cluster_name: xds_cluster
  cds_config:
    resource_api_version: V3
    ads: {}
  lds_config:
    resource_api_version: V3
    ads: {}

static_resources:
  clusters:
    - connect_timeout: 1s
      type: strict_dns
      http2_protocol_options: {}
      name: xds_cluster
      load_assignment:
        cluster_name: xds_cluster
        endpoints:
          - lb_endpoints:
              - endpoint:
                  address:
                    socket_address:
                      address: go-control-plane
                      port_value: 18000
```

that would resolve

im happy to open a ticket if what is happening is not expected, or update if it needs something more

File: docs/root/start/sandboxes/dynamic-configuration.rst
Comment (mattklein123) [reply docs/root/start/sandboxes/dynamic-configuration.rst]: Can you just remove ""host_rewrite_literal" ? It's not really needed for the example?

/wait

**Comment by phlax** [comment]
@mattklein123 ive finished updating - i think it should be ready for final review

not sure about https://github.com/envoyproxy/envoy/pull/13746#discussion_r513022567 but i reckon its beyond the scope of this PR

im happy to follow up

File: docs/root/start/sandboxes/dynamic-configuration.rst
Comment (phlax) [reply docs/root/start/sandboxes/dynamic-configuration.rst]: > Can you just remove ""host_rewrite_literal" ? It's not really needed for the example?

not without hacking go-control-plane

Comment (mattklein123) [reply docs/root/start/sandboxes/dynamic-configuration.rst]: Go control plane definitely does not force you to set this route config. How are you feeding the config into GCP?

Comment (phlax) [reply docs/root/start/sandboxes/dynamic-configuration.rst]: cool - i that seems to have worked - updated

**Review by mattklein123 — APPROVED** [review]
Awesome!!!

File: docs/root/start/sandboxes/_include/dynamic-configuration/response-config-cluster.json
Comment (phlax) [reply docs/root/start/sandboxes/_include/dynamic-configuration/response-config-cluster.json]: @mattklein123  lmk if you want me to follow up on this  in some way

Comment (mattklein123) [reply docs/root/start/sandboxes/_include/dynamic-configuration/response-config-cluster.json]: Ultimately we do want this to return v3 objects, but it's likely that the GCP samples need to be updated. cc @jyotima @kyessenov

Comment (phlax) [reply docs/root/start/sandboxes/_include/dynamic-configuration/response-config-cluster.json]: hmm, i reckon its not gcp thats returning this, its what you get from config above, when you do config_dump in  envoy - its the static part

Comment (mattklein123) [reply docs/root/start/sandboxes/_include/dynamic-configuration/response-config-cluster.json]: OK not sure what the issue is without digging in more detail. It's possible this is a config dump issue. I'm not sure.

Comment (phlax) [reply docs/root/start/sandboxes/_include/dynamic-configuration/response-config-cluster.json]: let me open a ticket - and we can thrash it out there...

Comment (jyotimahapatra) [reply docs/root/start/sandboxes/_include/dynamic-configuration/response-config-cluster.json]: @phlax Can you help me with the gcp samples you followed? From the initial pass looks like gcp example uses v3 https://github.com/envoyproxy/go-control-plane/blob/master/internal/example/README.md

Comment (phlax) [reply docs/root/start/sandboxes/_include/dynamic-configuration/response-config-cluster.json]: @jyotimahapatra i think its not a problem with go-control-plane

i opened a ticket here #13817 to raise it

**Pull request merged**
